### ISSUE TITLE
Feature: support syntax to drop value of matched token

### ISSUE THREAD

**Issue by conartist6** [kach/nearley#592]
I have a fairly complicated line:
```ne
Frame -> At FrameFragments _  LB "as" _ FrameFragments RB _ LP FrameFragments CN Number CN Number RP
{% (d) => ... %}
```

One of the most annoying things about working with this line is that I'm continually having to fiddle with the `d` indexes. Is the filename fragment `d[12]` now? It breaks a lot. What I'd like to be able to do is specify some expressions (subgrammars? what is the correct word?) to have their results omitted from the results argument (`d`). If this were possible then my javascript could be:
```js
([ function, method, file, line, col ]) => ({ function, method, file, line, col })
```

I don't know what the syntax would be. Actually I have any idea what syntax nearley actually supports since there is no API documentation, only guides.

**Comment by KillyMXI** [comment]
Destructure `d` and drop unneeded parts:
```js
([,function,,,,,method,,,,file,,line,,col,]) => ({ function, method, file, line, col })
```

It's still fairly easy to miscount comma numbers though (I can't guarantee I didn't make a mistake above without testing).
Refactoring your long grammar rules and breaking them into shorter parts will be a good thing to do, even if you're not going to reuse the parts elsewhere.

**Comment by KillyMXI** [comment]
It never come to mind for me to question this.
In fact, I quite like the simplicity of 1:1 match of rule contents.
I wouldn't question myself to break a complex rule apart - it's just more convenient all around.

Anyway, at this point I think it is safe for you to start looking at different grammar-based parser generators. It doesn't seem like `nearley` will go anywhere.
Some of alternatives offer features such as naming (see `peggy`) or skipping (see `@thi.ng/parse`) of rule parts via special markup.

**Comment by KillyMXI** [comment]
*I'm working on some weird stuff lately and I figured nearley has fundamental limitations that make it not working for me either (only CFG and only a very limited way to discard alternative branches).
So I'm about to publish a parser combinators package (because of course there is no parser combinators package that would work for me either).*

Why `_:?` tho?
By convention, `_` is an optional whitespace already, and `__` is a mandatory whitespace.

Ok, of course there are going to be edge cases but I think I'd deal with this line somehow or live with it (except for underscores - extra `:?` make it quite noisy).

**Comment by conartist6** [comment]
Ah right, I'll update that. `__`s sure would make it significantly more readable.

**Comment by conartist6** [comment]
Yeah named entries would be great, and even better if they were passed to the postprocessor in a separate argument. If the callback became `(tuple, dict, ...args) => value` it would be trivial to write `function dict(tuple, dict) { return dict; }`. In a lot of common cases this would eliminate the need for custom postprocessing.

**Comment by KillyMXI** [comment]
There are existing optional args for postprocessors after `d`.
Adding anything at the end would be unpretty and/or inconvenient.
Switching to options object pattern would be a breaking change.
Accepting an options object as an alternative to existing args would complicate the system.

So, adding stuff to `d` as an object would be least invasive and also pretty natural - just string keys in addition to numeric indices.

**Comment by conartist6** [comment]
Yes I'm proposing a breaking change -- just shifting the current optional callback args right. There doesn't seem to be any reason not to propose a breaking solution as likely it could only happen in a fork.

**Comment by KillyMXI** [comment]
Good point.

**Comment by KillyMXI** [comment]
Although a sensible AST would require at least one extra property not parsed from the input - node type.

Trying to add that automatically as well will start to look like a system optimized for a single purpose. If I were designing a parser optimized for AST generation I might've done the whole thing differently.

Keeping things general purpose as they are while optimizing for a particular single goal in mind might lead to weird compromises.
